hca/dss/__init__.py

```diff
@@ -114,8 +116,44 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
 
             logger.info("File %s: Retrieving...", filename)
             file_path = os.path.join(walking_dir, filename_base)
-            self._download_file(file_uuid, file_["sha256"], file_path, replica, file_['size'],
+            self._download_and_link_to_filestore(file_path,
+                                                 file_uuid,
+                                                 file_["sha256"],
+                                                 replica,
+                                                 file_['size'],
+                                                 file_version=file_version,
+                                                 num_retries=num_retries,
+                                                 min_delay_seconds=min_delay_seconds)
+
+    def _download_and_link_to_filestore(self, file_path, file_uuid, sha, replica, size, file_version, num_retries, min_delay_seconds):
+        file_store_path = self._download_to_filestore(file_uuid, sha, replica, size,
+                                                      file_version=file_version,
+                                                      num_retries=num_retries,
+                                                      min_delay_seconds=min_delay_seconds)
+        if sys.version_info < (3,) and platform.system() == 'Windows':
+            # Hardlinks not supported with Python 2 on Windows
+            raise ValueError('Upgrade Python or use a real OS')
+        os.link(file_store_path, file_path)
+
+    def _download_to_filestore(self,
```

Comment:
At this point it might make sense to lift these file-related arguments into an actual File object. This would also increase the expressiveness of the relationship between future and row by converting it into a relation between file and future.

The corresponding refactoring is https://refactoring.com/catalog/introduceParameterObject.html

Reply:
I like this idea.